v/vm.c

```diff
@@ -107,6 +117,57 @@ void printhex(uint64_t x)
   cputstring(str);
 }
 
+// for testing purpose only alloc
+static pte_t alloc()
+{
+  if (freelist_head == 0) {
+    return OUT_OF_MEMORY;
+  }
+  pte_t res = freelist_head->addr;
+  if (freelist_head != freelist_tail)
+    freelist_head = freelist_head->next;
+  else
+    freelist_head = freelist_tail = 0;
+  return res;
+}
+
+static uint64_t map(uint64_t vaddr, uint64_t paddr)
+{
+  // local variables
+  pte_t first_pte = pt[0][VPN2(vaddr)];
+  pte_t* second_pte_ptr;
+  pte_t second_pte;
+  pte_t* third_pte_ptr;
+  pte_t third_pte;
+
+  // first level                                                                                                        
+  if (first_pte & PTE_V) {
+    second_pte_ptr = ((pte_t*) (((first_pte & PTE_PPN) << PTE_PPN_OFFST) | (VPN1(vaddr) << PTE_OFF)));
+    second_pte = *second_pte_ptr;
+  } else {
+    // creating new page                                                                                               
+    // get a new page from global ppgdir                                                                                
+    uint64_t newPage = alloc();
+    pt[0][VPN2(vaddr)] = ((newPage & ~0xfff) >> PTE_PPN_OFFST) | PTE_V;
```

Comment:
the newpage's address should be asserted to be aligned with 4K pages instead of clearing the lower 12 bits.